paper-checkbox-light

bicknellr opened the pull request.

bicknellr commented:
I'll remove the validation tests soon since it no longer aplies; the line-height test failing was unexpected though.

notwaldorf commented:
@bicknellr Hmmm in `iron-input` I just did the validation manually.... you can add a TODO for that,  but I think this is functionality we want to keep.

notwaldorf commented on demo/paper-checkbox-light.html:
```diff
@@ -0,0 +1,91 @@
+      <h3>Checkboxes can be checked or unchecked, or disabled entirely</h3>
+      <demo-snippet class="centered-demo">
+        <template>
+          <paper-checkbox-light></paper-checkbox-light>
```
I would also add a demo for "what should you do for a label"

notwaldorf replied:
Actually, I would also bring back the subtitle examples, but with flexbox or whatever you need. Maybe just have a section "Checkboxes with labels", and do a bunch of examples? Like single line, multi-line-align-top, multi-line-align-center?

bicknellr replied:
Ok, I added them back.

notwaldorf commented on paper-checkbox-light.html:
```diff
@@ -0,0 +1,263 @@
+      }
+
+      :host-context([dir="rtl"]) paper-ripple {
+        right: calc(0px - (2.66 * var(--calculated-paper-checkbox-size) - var(--calculated-paper-checkbox-size)) / 2);
```
you double checked the ink looks sane, right?

bicknellr replied:
Yeah; since #130 isn't in, it still suffers from the clipping thing but I figure that should go in #130 itself or a later PR.

notwaldorf replied:
Oh SG. I just mean, "with RTL, it's where we expect it, not like 27 pixels to the right or something"

bicknellr replied:
Yeah, looks ok in RTL. Although paper-checkbox (regular and light) both do this weird thing where the checkmark expands from the top of the box - I bet #142 will fix this.

notwaldorf commented on paper-checkbox-light.html:
```diff
@@ -0,0 +1,263 @@
+      }
+    </style>
+
+    <div id="checkbox">
```
🙌

notwaldorf commented on paper-checkbox-light.html:
```diff
@@ -0,0 +1,263 @@
+<!--
+Material design: [Checkbox](https://www.google.com/design/spec/components/selection-controls.html#selection-controls-checkbox)
+
+`paper-checkbox-light` is a button that can be either checked or unchecked.  User
```
Pls add more details about what the difference between this and `paper-checkbox` is (i.e. faster/less features). People who wouldn't have been part of the offline discussions will look at this and be confused

bicknellr replied:
Added a better description.

notwaldorf commented on paper-checkbox-light.html:
```diff
@@ -0,0 +1,263 @@
+      },
+
+      attached: function() {
+        this.setAttribute('aria-checked', !!this.checked);
```
I think for the falsy case it needs to be `this.removeAttribute('aria-checked')`. Same for disabled.

bicknellr replied:
I think aria-disabled is ok to be explicitly 'false' - 'false' just happens to be the default too: https://www.w3.org/TR/wai-aria-1.1/#aria-disabled

According to this, if aria-checked is undefined, that indicates that the element isn't able to be checked: https://www.w3.org/TR/wai-aria-1.1/#aria-checked

notwaldorf commented on paper-checkbox-light.html:
```diff
@@ -0,0 +1,263 @@
+      },
+
+      _checkedChanged: function(checked) {
+        this.setAttribute('aria-checked', this.checked);
```
I think for the falsy case it needs to be `this.removeAttribute('aria-checked')`. Same for disabled.

bicknellr replied:
([comment in other thread](https://github.com/PolymerElements/paper-checkbox/pull/145#discussion_r73958983))

notwaldorf commented on paper-checkbox-light.html:
```diff
@@ -0,0 +1,263 @@
+      },
+
+      _onKeypress: function(event) {
+        if (event.key === ' ' || event.code === 'Space' || event.keyCode === 32) {
```
Does this work on all the browsers, old and new? (see: IE10)
What about `enter`?

bicknellr replied:
IE10, IE11, Edge, as well as current Firefox and Safari work with space. ARIA Authoring Practices doesn't say anything about enter needing to work, so I didn't include it. But enter does work for the previous paper-checkbox so I'll add it back.

bicknellr replied:
Enter works now.

notwaldorf commented on test/paper-checkbox-light.html:
```diff
@@ -0,0 +1,207 @@
+  <link rel="import" href="../paper-checkbox-light.html">
+
+  <style is="custom-style">
+    paper-checkbox-light.no-label-spacing {
```
Not a real property anymore

bicknellr replied:
removed

notwaldorf commented on test/paper-checkbox-light.html:
```diff
@@ -0,0 +1,207 @@
+    </template>
+  </test-fixture>
+
+  <test-fixture id="WithLabel">
```
Not a real test

bicknellr replied:
removed along with related tests

bicknellr commented:
`paper-checkbox-light` needs to include the ink sizing stuff from #135; hopefully that won't slow it down much.

bicknellr commented:
Also, `noink` doesn't work.

bicknellr commented:
Added `noink` support and fixed some associated keyboard vs. pointer ripple stuff. Also, I copied over the ink sizing code from #135 and delayed it until the ripple is first created. That would probably be ok to put in regular paper-checkbox also, but I'll save that for a later PR.

bicknellr commented:
As usual, Firefox and Selenium have reacted violently; everything seems ok locally though.

notwaldorf commented on paper-checkbox-light.html:
```diff
@@ -0,0 +1,327 @@
+       * Focus events occuring less than this value (in milliseconds) after a
+       * 'down' event are considered to be the result of a pointer.
+       */
+      __POINTER_FOCUS_WINDOW: 30,
```
Uhmmmm is this the alternative to `receivedFocusFromKeyboard`? Does it fix the problem where switching tabs focuses the element? Should we port this to `IronControlState` as well?

/cc @cdata

bicknellr replied:
> Uhmmmm is this the alternative to `receivedFocusFromKeyboard`?

Yeah, the behavior seems pretty much the same (as far as I can tell) except paper-checkbox-light also ripples when toggling with the keyboard after you initially focused it with a pointer. At some point I'd also like to make the persistent ripple come back when you toggle with the keyboard but I don't think it's strictly necessary for this PR, given regular paper-checkbox doesn't do this either.

> Does it fix the problem where switching tabs focuses the element?

Shouldn't it receive focus when you switch back to the page?

bicknellr commented:
The builds are now lying to us... They failed, even though it's just the usual Firefox + Selenium and Sauce Connect timeout problems, but they're reporting as having passed.

notwaldorf commented:
@bicknellr Yeah, i merged a PR yesterday that was all green, but the tests were legit broken :sob:

bicknellr commented:
Note to self: https://github.com/PolymerElements/paper-checkbox/issues/149
Normal `paper-checkbox` also needs to have ink size calculation delayed until focus or down.

bicknellr commented:
Note to self: https://github.com/PolymerElements/paper-checkbox/issues/149
Normal `paper-checkbox` also needs to have ink size calculation delayed until focus or down.

bicknellr commented:
ping @notwaldorf

notwaldorf commented on paper-checkbox-light.html:
```diff
@@ -0,0 +1,327 @@
+        var isSpace = event.key === ' ' || event.code === 'Space' || event.keyCode === 32;
+        var isEnter = event.key === 'Enter' || event.code === 'Enter' || event.keyCode === 13;
+        if (isSpace || isEnter) {
+          event.preventDefault();
```
Do you need `stopImmediatePropagation` too?

bicknellr replied:
I don't think so; `preventDefault` seems to be enough to stop scrolling.

notwaldorf commented on paper-checkbox-light.html:
```diff
@@ -0,0 +1,327 @@
+          observer: '_disabledChanged'
+        },
+
+        noink: {
```
Did you test that `noink` works in the demo?

bicknellr replied:
Wow, yup, this didn't work at all! I left it in a branch and never merged... Fixed now though.

notwaldorf commented on test/paper-checkbox-light.html:
```diff
@@ -0,0 +1,208 @@
+
+      test('checkbox respects the user set aria-label', function() {
+        var c = fixture('AriaLabel');
+        assert.isTrue(c.getAttribute('aria-label') == "Batman");
```
nit: should use `'` not `"`

bicknellr replied:
fixed

notwaldorf commented:
minor comments left. If the tests are green and the demos look ok, then LGTM

bicknellr commented on paper-checkbox-light.html:
```diff
@@ -0,0 +1,329 @@
+      attached: function() {
+        Polymer.RenderStatus.afterNextRender(this, function() {
+          if (!this._listenersAttached) {
+            this.listen(this, 'tap', '_onTap');
```
I moved these here from `listeners`.

bicknellr commented on paper-checkbox-light.html:
```diff
@@ -0,0 +1,329 @@
+      _disabledChanged: function(disabled) {
+        this.setAttribute('aria-disabled', disabled);
+        this.style.pointerEvents = disabled ? 'none' : '';
+        this.tabIndex = disabled ? -1 : 0;
```
The two lines above are new.

bicknellr commented:
I added https://github.com/PolymerElements/paper-checkbox/compare/3a14255...paper-checkbox-light.

bicknellr commented:
Ok, figured out why the tests were failing in some older browsers: I had used method syntax in an object literal.

bicknellr commented:
@notwaldorf mind taking another look?

LarsDenBakker commented:
Running into some performance issues with paper-checkbox, what's the status of this?
